02_bubble/marksomething/main_test.go

```diff
@@ -0,0 +1,96 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+	"testing"
+)
+
+func Test(t *testing.T) {
+
+	algorithms := []struct {
+		desc string
+		fn   func([]int) []int
+	}{
+		{"Bubble", bubble},
+		{"Insertion", insertion},
+		{"quick", quick},
+	}
+
+	testCases := []struct {
+		desc     string
+		arg      []int
+		expected []int
+	}{
+		{
+			desc:     "Basic",
+			arg:      []int{3, 2, 1, 5, 3, 2, 1, 7, 6},
+			expected: []int{1, 1, 2, 2, 3, 3, 5, 6, 7},
+		},
+		{
+			desc:     "NegativeValues",
+			arg:      []int{3, -2, 1, 5, -3, 2, 1, -7},
+			expected: []int{-7, -3, -2, 1, 1, 2, 3, 5},
+		},
+		{
+			desc:     "SingleEntry",
+			arg:      []int{1},
+			expected: []int{1},
+		},
+		{
+			desc:     "Empty",
+			arg:      []int{},
+			expected: []int{},
+		},
+	}
+	for _, algorithm := range algorithms {
+		for _, tC := range testCases {
+			testName := fmt.Sprintf("%v-%v", algorithm.desc, tC.desc)
+			alg := algorithm
+			testCase := tC
+			t.Run(testName, func(t *testing.T) {
+				actual := alg.fn(testCase.arg)
+				expected := testCase.expected
+				if !reflect.DeepEqual(actual, expected) {
+					t.Errorf("Actual (%q) does not equal Expected (%q)", actual, expected)
+				}
+
+			})
+		}
+	}
+}
+
+func TestMainOutput(t *testing.T) {
+	var buf bytes.Buffer
+	out = &buf
+
+	main()
+
+	actual := buf.String()
+	expected := "[1 2 3 5]\n[1 2 3 5]\n[1 2 3 5]\n"
+	if actual != expected {
+		t.Errorf("Actual (%v) does not equal Expected (%v)", actual, expected)
+	}
+}
+
+func testNoMutation(t *testing.T, alg func([]int) []int) {
+	actual := []int{3, 1, 2, 7, -5, 8}
+	expected := []int{3, 1, 2, 7, -5, 8}
+	alg(actual)
+
+	if !reflect.DeepEqual(actual, expected) {
+		t.Errorf("Actual (%q) does not equal Expected (%q)", actual, expected)
+	}
+}
+
+func TestBubbleNoMutation(t *testing.T) {
```

Comment:
Why didn't you wrap the sort functions in their own struct like you did before? thought that was really neat....
you could declare the cars inside the algo/func loop to ensure re-initialisation.

Reply:
Yeah, i got a little lazy there, have updated